### ISSUE TITLE
Is predict(..., pred_contrib=True) thread safe?

### ISSUE THREAD

**Issue by zyxue** [microsoft/LightGBM#5482]
## Description

I'm using the model behind a gRCP prediction service. The service predicts one example at a time.

I can reproduce the bug locally like

```python
data = [df.loc[:1][model.feature_name()]] * 1_000

def _predict(df_one_row):
    return model.predict(df_one_row, pred_contrib=True)

with ThreadPoolExecutor(max_workers=32) as exc:
    exc.map(_predict, data)
```

The error is like

```
free(): invalid next size (normal)
```
or 
```
double free or corruption (!prev)
```
or 
```
malloc(): corrupted top size
```

depending on different runs.

* If I set `max_workers=1`, the error seems to go away.
* If I remove `pred_contrib=True`, the error seems to go away, so it looks `predict` alone is thread-safe, consistent with https://github.com/microsoft/LightGBM/issues/666

## Reproducible example

I can produce some diff error, but also related to threading using the following code:

```python
from concurrent.futures import ThreadPoolExecutor

import sklearn.datasets
import lightgbm

df = (
    sklearn.datasets.load_iris(as_frame=True)["frame"]
    .sample(99, random_state=123)
    .rename(
        columns={
            "sepal length (cm)": "sepal_length",
            "sepal width (cm)": "sepal_width",
            "petal length (cm)": "petal_length",
            "petal width (cm)": "petal_width",
        }
    )
    .assign(sepal_length_cat=lambda df: (df.sepal_length > 1).astype(str).astype('category'))
    .reset_index(drop=True)
)

X, y = df.drop(columns="target"), df["target"]

regressor = lightgbm.LGBMRegressor(n_estimators=100, max_depth=7, objective="mse").fit(
    X, y
)

regressor.fit(X, y)

model = regressor.booster_

print(f'{df.dtypes=:}')

data = [df.loc[:1][model.feature_name()]] * 10_000


def _predict(df_one_row):
    return model.predict(df_one_row, pred_contrib=True)


with ThreadPoolExecutor(max_workers=32) as exc:
    exc.map(_predict, data)
```

The error is like

```
corrupted size vs. prev_size
corrupted size vs. prev_size
```
or 
```
python3: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
```

## Environment info

LightGBM version or commit hash:

lightgbm==3.3.2

Command(s) you used to install LightGBM

I'm using lightgbm inside a monorepo with bazel, but I think under the hood it's equivalent to `python -m pip install lightgbm`

**Comment by shuttie** [comment]
Seems to be an issue related to the `pred_contrib=True` handling in the C library itself: in the `lightgbm4j` library we have an issue with concurrent prediction crashing the whole JVM process due to the call not being thread-safe. See issue https://github.com/metarank/lightgbm4j/issues/88 for details and a reproducer.

**Comment by AndreyOrb** [comment]
I found a bug in Predictor related to incorrect usage of omp_get_thread_num.
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L94

When working in multi-threaded mode, different threads can return tid=0.
https://stackoverflow.com/questions/68484180/is-it-safe-to-use-omp-get-thread-num-to-index-a-global-vector
https://stackoverflow.com/questions/4087852/omp-set-num-threads-always-returns-0-and-im-unable-to-get-thread-num-with-omp-ge
https://stackoverflow.com/questions/43952078/openmp-always-working-on-the-same-thread

So, different threads (predict_fun_ lambdas) can write to the same memory block. 
There's another issue mentioning the same problem: https://github.com/microsoft/LightGBM/issues/3751

I think the same problem is in all 4 prediction types.

**Comment by jameslamb** [comment]
Thanks @AndreyOrb . Are you interested in proposing a fix? We'd welcome it!

It is probably worth looking at these other places, to see if they're safe:

```shell
git grep omp_get_thread_num
```

**Comment by AndreyOrb** [comment]
I have reviewed the issue carefully.
I actually identified 2 problems.

1) Improper Predictor lifecycle management:

    For each Predict request, there is a Predictor that is being created in
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/c_api.cpp#L463

    The Predictor class has predict_buf_, used for storing features:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L297

    Its constructor resizes this internal buffer according to number of features:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L66

    After the resize, a predictor function is constructed. It captures the address of the predict_buf_ vector:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L92

   All good, but when a Predictor object is passed out, a temporary object is created. Then, move semantics (if available) or copy semantics are used to initialize the predictor variable:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/c_api.cpp#L490

   In both cases, the original address is invalidated. When a function is started, it's looking at the predict_buf_ address that is not valid anymore:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L95

   I have tried refactoring the code to work with "move" operator, but again, it des not help because the address gets destroyed.
   Because of the location of the function creation (inside the constructor), the only way is to preserve the temporary Predictor object.
   So, I replaced the CreatePredictor's return type to std::shared_ptr<Predictor>. That solved the problem.

2) There is a unique Boosting object that is essentially a model.
    Then, again, in each Predict request, a new Predictor object is created.
    In its constructor, the GBDT's InitPredict is being called:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/application/predictor.hpp#L61

    The GBDT object is the Boosting itself, as of my understanding. It is unique for each model.
    InitPredict changes GBDT's internal attributes. So, running 2 different concurrent Predict requests are not thread-safe.
    To overcome the issue, there's a lock:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/c_api.cpp#L489
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/c_api.cpp#L57

    But, this lock is probably does not lock as expected, because I end up passing is in several threads:
![Image](https://github.com/user-attachments/assets/249891e4-6b46-42bd-907b-2be1e6c82e0e)

    Anyway, it's fine in my case, because all of my Predict requests have the same parameters (int start_iteration, int num_iteration, bool is_pred_contrib).
    The problem is here:
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/src/boosting/gbdt.h#L436

    This code runs RecomputeMaxDepth on each model's tree. When I hit this line in several threads, I get corrupted memory and all the application crashes. Moreover, I don't think this code should recompute the max tree depth in each contrib request. Please correct me, if I'm wrong. Maybe in training there's such a need.

    I've added a lock and proper condition (double-checked locking) to recompute the max tree depth only once.
    That has helped. I tested the fixed code by running 200k requests that run Predict (Regular & Contrib) in 20 parallel threads. No crashes.
    Will send the PR shortly after.
